src/graphs/SMGEdgeHasValueFilter.cc

```diff
@@ -0,0 +1,84 @@
+/*
+    Created by Viktor Malik on 19.2.2016.
+*/
+
+#include "SMGEdgeHasValueFilter.hh"
+#include "graphs/SMG.hh"
+
+namespace smg {
+
+SMGEdgeHasValueFilter::SMGEdgeHasValueFilter()
+    : object_(nullptr), value_(SMGValue::GetInvalidValue()), value_complement_(false), offset_(-1),
+      type_(SMGCType::GetInvalidType()) {
+}
+
+SMGEdgeHasValueFilter SMGEdgeHasValueFilter::ObjectFilter(const SMGObjectPtr& object) {
+  SMGEdgeHasValueFilter filter;
+  filter.FilterByObject(object);
+  return filter;
+}
+const SMGEdgeHasValueFilter& SMGEdgeHasValueFilter::FilterByObject(const SMGObjectPtr object) {
+  object_ = object;
+  return *this;
+}
+const SMGEdgeHasValueFilter& SMGEdgeHasValueFilter::FilterHavingValue(const SMGValue value) {
+  value_ = value;
+  value_complement_ = false;
+  return *this;
+}
+const SMGEdgeHasValueFilter& SMGEdgeHasValueFilter::FilterNotHavingValue(const SMGValue value) {
+  value_ = value;
+  value_complement_ = true;
+  return *this;
+}
+
+const SMGEdgeHasValueFilter& SMGEdgeHasValueFilter::FilterAtOffset(const long offset) {
+  offset_ = offset;
+  return *this;
+}
+
+const SMGEdgeHasValueFilter& SMGEdgeHasValueFilter::FilterByType(const SMGCType& type) {
+  type_ = type;
+  return *this;
+}
+
+bool SMGEdgeHasValueFilter::HoldsFor(const SMGEdgeHasValue& edge) const {
+  if (object_ && object_ != edge.GetObject()) {
+    return false;
+  }
+
+  if (value_ != SMGValue::GetInvalidValue()) {
+    if (value_complement_ && value_ == edge.GetValue()) {
+      return false;
+    } else if (!value_complement_ && value_ != edge.GetValue()) {
```

Comment:
I am not addressing efficiency here, not so much. I find the `else ` being redundant.
IF (abc)
   return;
ELSE IF (def)
   return;

Also, in other parts of the library, the same approach is used without the redundant `else`.